File: vol2/ch20/LeetCode752.cpp
struct QNode {                      //优先队列结点类型
  string s;       //数字字符串
      int f,g,h;    //启发式函数值
      bool operator<(const QNode &s) const {
          return f>s.f;                   //f越小越优先出队
      }
};
class Solution {
public:
  int openLock(vector<string>& deadends, string target) {
      if (target=="0000") return 0;
          unordered_set<string> dead(deadends.begin(),deadends.end());
          if (dead.count("0000")==1) return -1;
          priority_queue<QNode> pqu;
          QNode e,e1;
          e.s="0000";e.g=0;
          e.h=geth(e.s,target);
          e.f=e.g+e.h;
          pqu.push(e);
          unordered_set<string> visited={"0000"};
          while (!pqu.empty()) {
              e=pqu.top();pqu.pop();
              string s=e.s;
              for (string ns:getnexts(s)) {
                  if (visited.count(ns)==0 && dead.count(ns)==0) {
                      if (ns==target) return e.g+1;           //找到目标
                      e1.s=ns; e1.g=e.g+1;
                      e1.h=geth(ns,target);
                      e1.f=e1.g+e1.h;
                      pqu.push(e1);
                      visited.insert(ns);
                  }
              }
          }
          return -1;
      }
      char prev(char x) {                 //x逆时针方向旋转
          return x=='0'?'9':x-1;
      }
      char succ(char x) {                 //x顺时针方向旋转
          return x=='9'?'0':x+1;
      }
    vector<string> getnexts(string& s) {      //枚举s通过一次旋转得到的数字串
          vector<string> ret;
          for (int i=0;i<4;i++) {             //旋转每个数字符
              char c=s[i];
              s[i]=prev(c);               //逆时针方向旋转s[i]
              ret.push_back(s);               //旋转结果存入ret
              s[i]=succ(c);               //顺时针方向旋转s[i]
              ret.push_back(s);               //旋转结果存入ret
              s[i]=c;
          }
          return ret;
      }
   int geth(string& s,string& target) {       //计算启发函数值
      int h=0;
          for (int i=0;i<4;i++) {
              int dist=abs(int(s[i])-int(target[i]));
              h+=min(dist,10-dist);
          }
          return h;
      }
};

File: vol2/ch20/LeetCode752.py
class QNode:                #优先队列结点类型
    s=""                    #数字字符串
    f,g,h=0,0,0             #启发式函数值
    def __lt__(self,other):     #用于按f越小越优先出队
        if self.f<other.f:return True
        else:return False
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        if target=="0000":return 0
        dead=set()
        for e in deadends:dead.add(e)
        if "0000" in dead:return -1
        pqu=[]
        e=QNode()
        e.s="0000";e.g=0
        e.h=self.geth(e.s,target)
        e.f=e.g+e.h
        heapq.heappush(pqu,e)
        visited=set()
        visited.add("0000")
        while pqu:
            e=heapq.heappop(pqu)          #出队e
            s=e.s
            for ns in self.getnexts(s):
                if ns in visited:continue
                if ns in dead:continue
                if ns==target:return e.g+1      #找到目标
                e1=QNode()
                e1.s=ns; e1.g=e.g+1
                e1.h=self.geth(ns,target)
                e1.f=e1.g+e1.h
                heapq.heappush(pqu,e1)
                visited.add(ns)
        return -1
    def prev(self,x):                #x逆时针方向旋转
        if x=='0':return '9'
        else:return chr(ord(x)-1)
    def succ(self,x):               #顺时针方向旋转
        if x=='9':return '0'
        else:return chr(ord(x)+1)
    def getnexts(self,s):    #枚举s通过一次旋转得到的数字串
        ret=[]
        t=list(s)							#将s转换为字符列表t
        for i in range(0,4):         #旋转每个数字符
            c=t[i]
            t[i]=self.prev(c);               #逆时针方向旋转t[i]
            ret.append(''.join(t))      #旋转结果存入ret
            t[i]=self.succ(c);             #顺时针方向旋转t[i]
            ret.append(''.join(t))             #旋转结果存入ret
            t[i]=c
        return ret
    def geth(self,s,target):     #计算启发函数值
        h=0
        for i in range(0,4):
            dist=abs(int(s[i])-int(target[i]))
            h+=min(dist,10-dist)
        return h
